File: backend/microservice1/src/main/java/tn/esprit/microservice1/services/IntelligentCourseService.java
package tn.esprit.microservice1.services;

import org.springframework.stereotype.Service;
import org.springframework.beans.factory.annotation.Autowired;
import tn.esprit.microservice1.entities.*;
import tn.esprit.microservice1.repositories.CourseRepository;
import java.util.*;

@Service
public class IntelligentCourseService {
    
    @Autowired
    private CourseRepository courseRepository;

    /**
     * Generates personalized course recommendations based on user's learning history and preferences
     */
    public List<Course> getPersonalizedRecommendations(Long userId, int limit) {
        // Fetch user preferences and learning history
        List<Course> recommendedCourses = courseRepository.findTopNByPreferences(userId, limit);
        return recommendedCourses;
    }

    /**
     * Analyzes course content and generates AI tags and summary
     */
    public void enrichCourseWithAI(Course course) {
        // Generate AI tags based on course content
        List<String> aiTags = generateAITags(course);
        course.setAiGeneratedTags(aiTags);

        // Generate AI summary
        String aiSummary = generateAISummary(course);
        course.setAiGeneratedSummary(aiSummary);

        // Calculate difficulty score
        Double difficultyScore = calculateDifficultyScore(course);
        course.setDifficultyScore(difficultyScore);

        courseRepository.save(course);
    }

    /**
     * Generates adaptive learning path based on student's progress
     */
    public List<LearningObjective> generateAdaptiveLearningPath(Long userId, Long courseId) {
        // Analyze completed courses and suggest next learning objectives
        List<LearningObjective> objectives = new ArrayList<>();
        // Logic to determine next objectives based on user progress
        return objectives;
    }

    /**
     * Identifies knowledge gaps based on assessment results
     */
    public Map<String, Double> identifyKnowledgeGaps(Long userId, Long courseId) {
        // Analyze assessment results to identify gaps
        Map<String, Double> knowledgeGaps = new HashMap<>();
        // Logic to analyze results and populate knowledgeGaps
        return knowledgeGaps;
    }

    /**
     * Generates personalized practice questions based on identified weak areas
     */
    public List<String> generatePersonalizedQuestions(Long userId, Long courseId) {
        // Generate questions based on identified weak areas
        List<String> questions = new ArrayList<>();
        // Logic to create questions
        return questions;
    }

    private List<String> generateAITags(Course course) {
        // TODO: Implement AI tag generation using NLP
        return Arrays.asList("ai-generated-tag-1", "ai-generated-tag-2");
    }

    private String generateAISummary(Course course) {
        // TODO: Implement AI summary generation using NLP
        return "AI-generated summary placeholder";
    }

    private Double calculateDifficultyScore(Course course) {
        // TODO: Implement difficulty score calculation
        return 0.5; // Placeholder score between 0 and 1
    }

    /**
     * Analyzes student engagement patterns
     */
    public Map<String, Object> analyzeEngagementPatterns(Long userId, Long courseId) {
        Map<String, Object> patterns = new HashMap<>();
        // Logic to analyze engagement patterns
        return patterns;
    }

    /**
     * Predicts student performance and completion likelihood
     */
    public Map<String, Double> predictPerformance(Long userId, Long courseId) {
        Map<String, Double> predictions = new HashMap<>();
        // Logic to predict performance based on historical data
        predictions.put("completionLikelihood", 0.75); // Example value
        predictions.put("expectedScore", 85.0); // Example value
        return predictions;
    }
} 

File: frontend/backoffice/src/app/course-management/services/intelligent-course.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Course, LearningObjective } from '../models/course.model';

@Injectable({
  providedIn: 'root'
})
export class IntelligentCourseService {
  private apiUrl = `${environment.apiUrl}/intelligent`;

  constructor(private http: HttpClient) {}

  /**
   * Get AI-powered course analytics
   */
  getCourseAnalytics(): Observable<any> {
    return this.http.get(`${this.apiUrl}/analytics`);
  }

  /**
   * Get personalized course recommendations
   */
  getPersonalizedRecommendations(userId: number, limit: number = 5): Observable<Course[]> {
    return this.http.get<Course[]>(`${this.apiUrl}/recommendations/${userId}?limit=${limit}`);
  }

  /**
   * Generate adaptive learning path using AI
   */
  getAdaptiveLearningPath(courseId: number, userId: number): Observable<LearningObjective[]> {
    return this.http.get<LearningObjective[]>(`${this.apiUrl}/courses/${courseId}/learning-path/${userId}`);
  }

  /**
   * Identify knowledge gaps using AI analysis
   */
  getKnowledgeGaps(courseId: number, userId: number): Observable<Map<string, number>> {
    return this.http.get<Map<string, number>>(`${this.apiUrl}/courses/${courseId}/knowledge-gaps/${userId}`);
  }

  /**
   * Get AI-generated practice questions
   */
  getPersonalizedQuestions(courseId: number, userId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/courses/${courseId}/practice-questions/${userId}`);
  }

  /**
   * Analyze student engagement patterns
   */
  getEngagementAnalysis(courseId: number, userId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/courses/${courseId}/engagement/${userId}`);
  }

  /**
   * Get AI predictions for student performance
   */
  getPredictedPerformance(courseId: number, userId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/courses/${courseId}/performance-prediction/${userId}`);
  }

  /**
   * Enrich course with AI-generated content
   */
  enrichCourseWithAI(courseId: number): Observable<void> {
    return this.http.post<void>(`${this.apiUrl}/courses/${courseId}/enrich`, {});
  }

  /**
   * Generate automated course content
   */
  generateAutomatedContent(courseId: number, parameters: any): Observable<any> {
    return this.http.post<any>(`${this.apiUrl}/courses/${courseId}/generate-content`, parameters);
  }

  /**
   * Get AI-powered content recommendations
   */
  getContentRecommendations(courseId: number, userId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/courses/${courseId}/content-recommendations/${userId}`);
  }

  /**
   * Analyze course difficulty and suggest optimizations
   */
  analyzeCourseComplexity(courseId: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/courses/${courseId}/complexity-analysis`);
  }

  /**
   * Generate personalized learning objectives
   */
  generatePersonalizedObjectives(courseId: number, userId: number): Observable<LearningObjective[]> {
    return this.http.get<LearningObjective[]>(`${this.apiUrl}/courses/${courseId}/personalized-objectives/${userId}`);
  }
} 